Kernel/Arch/riscv64/PCI/Initializer.cpp

```diff
@@ -4,17 +4,159 @@
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#include <Kernel/Arch/CPU.h>
+#include <Kernel/Boot/CommandLine.h>
+#include <Kernel/Bus/PCI/API.h>
+#include <Kernel/Bus/PCI/Access.h>
+#include <Kernel/Bus/PCI/Controller/MemoryBackedHostBridge.h>
 #include <Kernel/Bus/PCI/Initializer.h>
+#include <Kernel/FileSystem/SysFS/Subsystems/Bus/PCI/BusDirectory.h>
+#include <Kernel/Interrupts/GenericInterruptHandler.h>
 #include <Kernel/Library/Assertions.h>
+#include <Kernel/Memory/MemoryManager.h>
+#include <Userland/Libraries/LibDeviceTree/FlattenedDeviceTree.h>
 
 namespace Kernel::PCI {
 
 bool g_pci_access_io_probe_failed { false };
-bool g_pci_access_is_disabled_from_commandline { true };
+bool g_pci_access_is_disabled_from_commandline;
 
 void initialize()
 {
-    TODO_RISCV64();
+    g_pci_access_is_disabled_from_commandline = kernel_command_line().is_pci_disabled();
+    if (g_pci_access_is_disabled_from_commandline)
+        return;
+
+    new Access();
+
+    // https://github.com/devicetree-org/devicetree-specification/releases/download/v0.4/devicetree-specification-v0.4.pdf
+    // https://github.com/devicetree-org/dt-schema/blob/main/dtschema/schemas/pci/pci-bus-common.yaml
+    // https://github.com/devicetree-org/dt-schema/blob/main/dtschema/schemas/pci/pci-host-bridge.yaml
+
+    // The pci controllers are usually in /soc/pcie?@XXXXXXXX
+    // FIXME: They can also appear in the root node, or any simple-bus other than soc
+    auto const& device_tree = DeviceTree::get();
+
+    auto maybe_soc = device_tree.get_child("soc"sv);
+    if (!maybe_soc.has_value()) {
+        dmesgln("PCI: No `soc` node found in the device tree, PCI initialization will be skipped");
+        return;
+    }
+
+    auto const& soc = maybe_soc.value();
+
+    enum class ControllerCompatible {
+        CAM,
+        ECAM,
+    };
+
+    // These properties must be present
+    auto soc_address_cells = soc.get_property("#address-cells"sv).value().as<u32>();
+    [[maybe_unused]] auto soc_size_cells = soc.get_property("#size-cells"sv).value().as<u32>();
+
+    Optional<u32> domain_counter;
+    for (auto const& entry : soc.children()) {
+        if (!entry.key.starts_with("pci"sv))
+            continue;
+        auto const& node = entry.value;
+
+        if (auto device_type = node.get_property("device_type"sv); !device_type.has_value() || device_type.value().as_string() != "pci"sv) {
```

Comment:
`device_type` is deprecated (chapter 2.3.13), however the bindings say it should always be `pci`.
Not sure if we should check it here as it's deprecated.

Reply:
I think, if it is present we should check it, but if it is missing we can just ignore it